Add support to create external snapshots

phrdina opened the pull request:
Add support to create external snapshots. Currently the code is simple and all checks will be done by libvirt to figure out if creating snapshot is possible or not.

When creating external snapshot for running VM user needs to provide path where to store memory state. VirtManager will suggest paths based on disks location and the file name will be based on the VM name and snapshot name.

Creating external snapshot for offline VM:
![Screenshot_2024-03-04_12-17-08](https://github.com/virt-manager/virt-manager/assets/109641/d6799772-b6bc-4cbc-9957-f49b6348c683)

Creating external snapshot for running VM:
![Screenshot_2024-03-04_12-17-48](https://github.com/virt-manager/virt-manager/assets/109641/45772faa-0921-443d-a25e-b6460cd2604f)

crobinso reviewed (changes requested):
(this comment is about the last path)

Interesting. Why does libvirt start qemu? To use qemu to flatten the storage chain?

So VM is started as paused, CPUs are never started. But at least from virt-manager logs, libvirt is not firing off a domain lifecycle event. Do you know why not? Implicitly because CPUs are never started, or because sending the signal is explicitly suppressed for the snapshot delete code path?

If libvirt is doing some active suppression, maybe it should suppress the qemu-ga signal too?

Regardless, working around it in virt-manager is fine by me. better IMO to skip the obj.recache_from_event_loop call entirely: if VM isn't active, log a message like 'received agent signal but VM is inactive. skipping refresh' from the signal handler and return early. And add a comment that this triggers for the snapshot delete case.

FWIW it looks like on VM startup first signal we get is the agent DISCONNECTED signal too which seems weird, so maybe there's a bigger libvirt bug here anyways.

crobinso commented on virtManager/details/snapshots.py:
```diff
@@ -234,7 +302,7 @@ def _get_screenshot_data_for_save(self):
         ignore = asyncjob
 
-        self.vm.create_snapshot(xml, diskOnly=False)
+        self.vm.create_snapshot(xml, diskOnly=self._diskOnly)
```
There's a little bug here in that self._diskOnly=True can be carried over, if user creates a disk only external snapshot, and then next creates an internal snapshot. Drop `self._diskOnly` and just calculate it _create_new_snapshot and pass it to _do_create_snapshot.

Or adjust domain.py to take the xmlobj directly, and calculate it there, so this file doesn't need to care about the snapshot flags distinction.

phrdina replied:
Nice catch, will fix it!

crobinso commented:
Nice, glad it looks so simple. Must have been a lot of work getting that all going in libvirt (I'm out of the loop)

phrdina commented on tests/data/capabilities/kvm-x86_64.xml:
```diff
@@ -194,6 +194,7 @@
       <cpuselection/>
       <deviceboot/>
       <disksnapshot default='on' toggle='no'/>
+      <externalSnapshot/>
```
Nothing I know about. If that happens and we are able to detect it in libvirt from QEMU capabilities it should be simple to add.

phrdina commented on virtManager/details/snapshots.py:
```diff
@@ -108,9 +110,48 @@ def _cleanup(self):
+        model.clear()
+
+        for disk in self.vm.get_xmlobj().devices.disk:
+            dirpath = os.path.dirname(disk.get_source_path())
```
It could be interesting to users having VMs with multiple disks located in different places on the host system. I was considering adding code to figure out boot order and sort the paths based on that but eventually decided it's not worth the effort.

I'll probably changes the way how we tell libvirt to make external snapshot. For running VM adding path to memory state should be enough and for offline VM using disk-only is enough as well. So we will no longer hardcode any disk to the snapshot XML.

phrdina commented on ui/snapshotsnew.ui:
>     * Better IMO for mode to be a radio button group. Will let us add tooltips in the future, and make it more self documenting when modes aren't supported.

Good point, I'll change it.

>     * Do you think we should also change snapshot details window to show memory path and disk path? Feels a bit weird to ask for memory path but then not show it. But not sure how much people care about that data after the fact.

We could possibly add memory path to the details as we allow users to set their own path. For disks they technically can have any path as well but by default libvirt places the new overlay files in the same location as original disk and the new overlay file is part of the VM XML so I don't think there is a need to have disk paths in snapshot detail.

>     * This ain't a requirement but if you are motivated: memory field could be a single checkbox 'Auto' next to a text entry field. Auto selected by default, text field is disabled but visible, and as snapshot name is changed the memory field is updated to list the full memory path we will use. Uncheck auto to let the user manually enter a full memory path.

I was considering having full memory path there and change it if the snapshot name changes as well but was not sure how to handle it if user changes the file name. With the `auto` checkbox that should be easy and I like this suggestion.

phrdina replied:
> > ```
> > * Do you think we should also change snapshot details window to show memory path and disk path? Feels a bit weird to ask for memory path but then not show it. But not sure how much people care about that data after the fact.
> > ```
> 
> We could possibly add memory path to the details as we allow users to set their own path. For disks they technically can have any path as well but by default libvirt places the new overlay files in the same location as original disk and the new overlay file is part of the VM XML so I don't think there is a need to have disk paths in snapshot detail.

Eventually I've decided there is no need to show the path as there is nothing for users to do with the path. It would make sense to show it if we add also paths to all overlays but that would get complicated with multiple disks and snapshot branches and so on.

phrdina commented:
> (this comment is about the last path)
> 
> Interesting. Why does libvirt start qemu? To use qemu to flatten the storage chain?

Eventually we will use qemu-storage-daemon in libvirt and this issue should eventually go away and yes, qemu-img doesn't support block operations so we need to start qemu process to do `block-commit`.

> So VM is started as paused, CPUs are never started. But at least from virt-manager logs, libvirt is not firing off a domain lifecycle event. Do you know why not? Implicitly because CPUs are never started, or because sending the signal is explicitly suppressed for the snapshot delete code path?

It was supposed to be implementation details that the QEMU process is started and that specific event is directly passed through from QEMU and I did not realized something like this will happen.

> If libvirt is doing some active suppression, maybe it should suppress the qemu-ga signal too?

Agreed, should figure out some solution in libvirt as well, suppressing the signal or not presenting the VM as paused. But we will need the workaround in virt-manager as well for libvirt versions that will produce that event as it basically locks the VM until virt-manager is restarted or the connection is restarted.

> Regardless, working around it in virt-manager is fine by me. better IMO to skip the obj.recache_from_event_loop call entirely: if VM isn't active, log a message like 'received agent signal but VM is inactive. skipping refresh' from the signal handler and return early. And add a comment that this triggers for the snapshot delete case.

Will look into it.

> FWIW it looks like on VM startup first signal we get is the agent DISCONNECTED signal too which seems weird, so maybe there's a bigger libvirt bug here anyways.

Yeah it is strange, TBH I did not look into the details how exactly this works in QEMU and when exactly this specific event should happen, so I'll probably look into it to figure out the best approach to handle this.

Thanks for the review!

phrdina commented:
New version is pushed now and I've updated screenshots to reflect current state.

crobinso commented:
> > Regardless, working around it in virt-manager is fine by me. better IMO to skip the obj.recache_from_event_loop call entirely: if VM isn't active, log a message like 'received agent signal but VM is inactive. skipping refresh' from the signal handler and return early. And add a comment that this triggers for the snapshot delete case.
> 
> Will look into it.

Did you try this bit? Last patch seems the same. I'd prefer if we can avoid pushing changes to `libvirtobject.py` unless really necessary. The comment will help too

The only other bit is the diskonly= handling. It touches UI with `self._get_mode()` but we are inside a thread, which is dangerous. Pre-calc diskonly in _create_new_snapshot and pass it in. (it's possibly readonly UI access is ok in a thread but I don't want to risk it)

There's also a minor pylint whine: tests/test_capabilities.py:58: [E712] comparison to True should be 'if cond is True:' or 'if cond:'


If you fix those things feel free to push/merge

phrdina commented:
> > > Regardless, working around it in virt-manager is fine by me. better IMO to skip the obj.recache_from_event_loop call entirely: if VM isn't active, log a message like 'received agent signal but VM is inactive. skipping refresh' from the signal handler and return early. And add a comment that this triggers for the snapshot delete case.
> > 
> > 
> > Will look into it.
> 
> Did you try this bit? Last patch seems the same. I'd prefer if we can avoid pushing changes to `libvirtobject.py` unless really necessary. The comment will help too

Completely missed this when I was making sure all changes were done. Fixed now by ignoring the event for shutoff VMs.

> The only other bit is the diskonly= handling. It touches UI with `self._get_mode()` but we are inside a thread, which is dangerous. Pre-calc diskonly in _create_new_snapshot and pass it in. (it's possibly readonly UI access is ok in a thread but I don't want to risk it)

Fixed.

> There's also a minor pylint whine: tests/test_capabilities.py:58: [E712] comparison to True should be 'if cond is True:' or 'if cond:'

Fixed.

> If you fix those things feel free to push/merge
